core/lib/types/src/storage/writes/mod.rs

```diff
@@ -0,0 +1,499 @@
+use std::convert::TryInto;
+
+use crate::H256;
+use serde::{Deserialize, Serialize};
+use zksync_basic_types::{Address, U256};
+
+use self::compression::compress_with_best_strategy;
+
+mod compression;
+
+const COMPRESSION_VERSION_NUMBER: u8 = 1;
+const BYTES_PER_ENUMERATION_INDEX: u8 = 4;
+const STATE_DIFF_RECORD_SIZE: usize = 156;
+
+// 2 * 136 - the size that allows for two keccak rounds.
+pub const PADDED_ENCODED_STORAGE_DIFF_LEN_BYTES: usize = 272;
+
+/// In vm there are two types of writes Initial and Repeated. After the first write to the leaf,
+/// we assign an index to it and in the future we should use index instead of full key.
+/// It allows us to compress the data.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct InitialStorageWrite {
+    pub index: u64,
+    pub key: U256,
+    pub value: H256,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, Default, Eq, PartialEq)]
+pub struct RepeatedStorageWrite {
+    pub index: u64,
+    pub value: H256,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, Default, Eq, PartialEq)]
+pub struct StateDiffRecord {
+    /// address state diff occured at
+    pub address: Address,
+    /// storage slot key updated
+    pub key: U256,
+    /// derived_key == Blake2s(bytes32(address), key)
+    pub derived_key: [u8; 32],
+    /// index in tree of state diff
+    pub enumeration_index: u64,
+    /// previous value
+    pub initial_value: U256,
+    /// updated value
+    pub final_value: U256,
+}
```

Comment:
hmm so I think we have some structrs that represent this, let's think if we can reuse (`LogQuery` )

Reply:
i think it would be possible but it might not be the best idea given that most fields would not be used, we'd need to introduce some fields for this particular usage, and logically a StateDiff is the overall storage update (can be a combination of `LogQuery`)

Reply:
will quickly ping @perekopskiy here - what do you think, is a new struct justified here? I personally don't have a strong opionoin on that

Reply:
I support having a new struct